text/0000-layer-table.md

```diff
@@ -0,0 +1,89 @@
+[meta]: #meta
+- Name: Move layer types to new table in layer.toml
+- Start Date: 2021-01-05
+- Author(s): natalieparellano
+- RFC Pull Request: (leave blank)
+- CNB Pull Request: (leave blank)
+- CNB Issue: (leave blank)
+- Supersedes: N/A
+
+# Summary
+[summary]: #summary
+
+Introduces a new table in `<layers>/<layer>.toml` to enable Bash buildpacks to easily opt in to layer re-use.
+
+# Definitions
+[definitions]: #definitions
+
+# Motivation
+[motivation]: #motivation
+
+[RFC 0052](https://github.com/buildpacks/rfcs/blob/main/text/0052-opt-in-layer-caching.md) described a change (to be implemented in Buildpack API 0.6) whereby buildpacks must opt in to layer re-use - i.e., each `<layers>/<layer>.toml` will be restored with no `launch`, `build`, or `cache` flags set and buildpacks must modify the restored TOML to set these flags.
+
+The RFC assumed that Bash buildpacks would be able to easily opt in to layer re-use by appending lines to the TOML, e.g.:
+
+`echo "launch = true" >> layer.toml`
+
+However, because `launch`, `build`, and `cache` are top level keys, the above command would add a `launch` key to the `[metadata]` table instead of the top level.
+
+# What it is
+[what-it-is]: #what-it-is
+
+This RFC proposes moving the `launch`, `build`, and `cache` keys into a new `types` table in `<layers>/<layer>.toml`:
+
+```toml
+[types]
+launch = true
+build = false
+cache = false
+```
+
+A Bash buildpack could then write something like the following:
+
+```bash
+cat >> launch.toml <<EOL
+[types]
+launch = true
+EOL
+```
+
+This change should be implemented with Buildpack API 0.6 so that it goes hand-in-hand with the opt in layer re-use feature.
+
+- If a buildpack implementing Buildpack API less than 0.6 tries to write `<layers>/<layer>.toml` with a `[types]` table, the lifecycle should fail.
+- If a buildpack implementing Buildpack API 0.6 or above tries to write `<layers>/<layer>.toml` with a top level `launch`, `build`, or `cache` key, the lifecycle should fail.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- Why should we *not* do this? Having `launch`, `build`, and `cache` as top level keys potentially makes more sense in describing a layer.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- Bash buildpack authors could prepend to the TOML file.
+- What is the impact of not doing this? Potentially aggravating extra labor for Bash buildpack authors.
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- **Out of scope:** Since we are moving the `launch`, `build`, and `cache` keys, could they be re-named for further clarity?
+- Is `types` the right name for this table?
+- Would any changes be necessary to [`io.buildpacks.lifecycle.metadata`](https://github.com/buildpacks/spec/blob/main/platform.md#iobuildpackslifecyclemetadata-json)? This label includes `launch`, `build`, and `cache` keys for each referenced layer. If so, should the lifecycle standardize this label for buildpacks implementing different Buildpack APIs?
```

Comment:
Trying to circle back from the last discussion, did we mark this as out of scope or did we want to try to align label changes? @ekcasey

Reply:
I lean towards yes, let's standardize the label. But there is one caveat Once you build an image with this platform API it could have implications for a rebuild with an older lifecycle that doesn't understand the new format. If lifecycle downgrade is a sticking point for folks I would rather keep the label the same and move forward for now.

Reply:
I'm not sure I follow the need to change the label (this suggestion was more compelling when we were talking about changing the key names). The spec does require that analyzed.toml and `io.buildpacks.lifecycle.metadata` have the same schema, but the translation from <layer>.toml to analyzed.toml is more hand wavy; as it is, today <layer>.toml has `metadata` and the label has `data` to represent the same thing, so further translating <layer>.toml feels okay to me.

Reply:
Hmm... I may have been mistaken about what is `data` and what is `metadata` ...checking.

Reply:
I buy that argument. Let's leave it alone for now and we can update the label and `analyzed.toml` in the future if we wish.